src/trio/_subprocess.py

```diff
@@ -760,38 +763,45 @@ async def read_output(
-        try:
-            if input is not None:
-                assert proc.stdin is not None
-                nursery.start_soon(feed_input, proc.stdin)
-                proc.stdin = None
-                proc.stdio = None
-            if capture_stdout:
-                assert proc.stdout is not None
-                nursery.start_soon(read_output, proc.stdout, stdout_chunks)
-                proc.stdout = None
-                proc.stdio = None
-            if capture_stderr:
-                assert proc.stderr is not None
-                nursery.start_soon(read_output, proc.stderr, stderr_chunks)
-                proc.stderr = None
-            task_status.started(proc)
-            await proc.wait()
-        except BaseException:
-            with trio.CancelScope(shield=True):
-                killer_cscope = trio.CancelScope(shield=True)
-
-                async def killer() -> None:
-                    with killer_cscope:
-                        await deliver_cancel(proc)
-
-                nursery.start_soon(killer)
+    # Opening the process does not need to be inside the nursery, so we put it outside
+    # so any exceptions get directly seen by users.
+    # options needs a complex TypedDict. The overload error only occurs on Unix.
+    proc = await open_process(command, **options)  # type: ignore[arg-type, call-overload, unused-ignore]
+    try:
+        async with trio.open_nursery() as nursery:
+            try:
+                if input is not None:
+                    assert proc.stdin is not None
+                    nursery.start_soon(feed_input, proc.stdin)
+                    proc.stdin = None
+                    proc.stdio = None
+                if capture_stdout:
+                    assert proc.stdout is not None
+                    nursery.start_soon(read_output, proc.stdout, stdout_chunks)
+                    proc.stdout = None
+                    proc.stdio = None
+                if capture_stderr:
+                    assert proc.stderr is not None
+                    nursery.start_soon(read_output, proc.stderr, stderr_chunks)
+                    proc.stderr = None
+                task_status.started(proc)
                 await proc.wait()
-                killer_cscope.cancel()
-                raise
+            except BaseException:
+                with trio.CancelScope(shield=True):
+                    killer_cscope = trio.CancelScope(shield=True)
+
+                    async def killer() -> None:
+                        with killer_cscope:
+                            await deliver_cancel(proc)
+
+                    nursery.start_soon(killer)
+                    await proc.wait()
+                    killer_cscope.cancel()
+                    raise
+    except BaseExceptionGroup as exc:
+        if all(isinstance(e, Cancelled) for e in exc.exceptions):
+            raise exc
+        raise TrioInternalError("Error interacting with opened process") from exc
```

Comment:
I'd love some extra eyes on this particular change from people more familiar with the inner workings of trio. Maybe @richardsheridan ?

Reply:
I had thoughts on this. the first one is opening the process outside of the nursery, which you already did. the second one is that if `deliver_cancel` raises something other than `Cancelled`, it would be ideal to raise that error not in an exception group (or as a trio internal error.) As you might have seen in chat, there are several ways to accomplish this and none are very nice. Finally, I'm not sure it's valuable to convert pipe reader task error groups to a trio internal error. They may well be critical/fatal OS errors outside of trio's control. But since either the old or new way will be practically uncatchable, I don't have a strong preference.

Reply:
It's fairly easy to add a
```python
[ re-raise cancelled ]
if len(exc.exceptions) == 1:
  raise exc.exceptions from None
[ raise triointernalerror ]
```
clause to pass through single errors, and only resort to `TrioInternalError` (or something else) in group cases. But then we're causing users a headache if the same underlying exception can cause them to get two different exceptions depending on if a second exception occurred at the same time.
`TrioInternalError` does seem somewhat off, especially if we consider getting a single error not to be an internal error. But not sure what else to raise. RuntimeError? Or be upfront that we would raise a group in that case? Always wrap exceptions in a group? Just raise the first non-cancelled error, and have the group as a `__cause__`?

Reply:
sanity check: if I set the nursery in question here to `strict_exception_groups=False`, and then fudge `test_custom_deliver_cancel` so line 589 reads `assert False`, shouldn't I get a single unwrapped `AssertionError`? Right now I'm getting an EG

Reply:
which nursery is "the nursery in question", the one in test_subprocess L591, or the one in _subprocess line ~771?
If you're changing the former, you should get an EG, because it doesn't set the global contect of strict_exception_groups=False, so the one on 771 will use the default of True (in this PR).

Reply:
but passing a bad `deliver_cancel` is indeed a very trivial way to get an EG atm, hrm.